lib/spack/spack/solver/concretize.lp

```diff
@@ -382,6 +382,58 @@ external_conditions_hold(Package, LocalIndex) :-
 #defined external_spec_condition/4.
 #defined external_spec_condition/5.
 
+%-----------------------------------------------------------------------------
+% Config required semantics
+%-----------------------------------------------------------------------------
+
+% A package may have an associated "stanza": a set of specs -
+% one of which must be applied (but only one of them), that spec
+% is represented here with cfg_required(Package, Y), and may apply multiple
+% constraints (version, compiler, or variant settings).
+1 { cfg_required(Package, Y) : stanza_member(Y, X) } :- node(Package), required_stanza(Package, X).
+% asp.py must define required_stanzas
+% For each stanza, asp.py must define stanza_members for that stanza
+% For each stanza_member, asp.py must define cfg_*_required (the set of spec
+% constraints which apply for that stanza member)
+version(Package, X) :- cfg_required(Package, Y), cfg_version_required(Y, X).
+variant_value(Package, Variant, Value) :- cfg_required(Package, Y), cfg_variant_required(Y, Variant, Value).
+node_compiler(Package, X) :- cfg_required(Package, Y), cfg_compiler_required(Y, X).
+node_compiler_version(Package, C, V) :- cfg_required(Package, Y), cfg_compiler_version_required(Y, C, V).
+
+% TODO: don't warn if program is missing cfg_version_required
+% #defined cfg_version_required/1.
+
+% TODO: allow weighting the conditions; a condition should be preferred if it
+% appears first in a condition_group. Compared to other preference weights
+% this is less important, but presumably the user might want to exert some
+% level of control by ordering the preferences. This will probably involve
+% a #minimize statement like others at the end of this file
+
+% TODO: if the user provides a constraint on the command line, then it will
+% need to constrain the spec in a similar way (although in that case
+% there is no such thing as a stanza: there is just one spec to constrain by).
+% It looks like that is handled by asp._spec_clauses. That function would need
+% to be modified to cover this section though (since unlike there, the
+% constraints here are only applied if the associated stanza member is
+% selected).
+
+%-----------------------------------------------------------------------------
+% Generalization of prior section
+%-----------------------------------------------------------------------------
```

Comment:
`asp._spec_clauses` could be refactored to generate condition_groups or directly generate conditions (e.g. it would do this to apply spec constraints supplied from the command line).

Reply:
It turns out that a couple functions in `asp` (namely `condition` and `impose`) already use `_spec_clauses` to conditionally apply Spec constraints, so I can update this (and `asp.py`) to follow their lead.